Index: twisted/web2/dav/noneprops.py
===================================================================
--- twisted/web2/dav/noneprops.py	(revision 19773)
+++ twisted/web2/dav/noneprops.py	(working copy)
@@ -49,6 +49,9 @@
     def __init__(self, resource):
         pass
 
+    def cacheAllChildProperties(self):
+        pass
+
     def get(self, qname):
         raise HTTPError(StatusResponse(responsecode.NOT_FOUND, "No such property: {%s}%s" % qname))
 
@@ -60,8 +63,19 @@
         # non-existing property is not an error.
         pass
 
+    def deleteall(self, qname):
+        # RFC 2518 Section 12.13.1 says that removal of
+        # non-existing property is not an error.
+        pass
+
     def contains(self, qname):
         return False
 
     def list(self):
         return ()
+
+    def copy(self, props):
+        pass
+
+    def move(self, props):
+        pass
