Index: twisted/web2/dav/element/parser.py
===================================================================
--- twisted/web2/dav/element/parser.py	(revision 18545)
+++ twisted/web2/dav/element/parser.py	(working copy)
@@ -123,15 +123,10 @@
             for attr_name in attributes.getQNames():
                 attributes_dict[attr_name.encode("utf-8")] = attributes.getValueByQName(attr_name)
 
-        tag_namespace, tag_name = name
-
-        if (name not in elements_by_tag_name):
-            class UnknownElement (WebDAVUnknownElement):
-                namespace = tag_namespace
-                name      = tag_name
-            element_class = UnknownElement
-        else:
+        if name in elements_by_tag_name:
             element_class = elements_by_tag_name[name]
+        else:
+            element_class = WebDAVUnknownElement
 
         self.stack.append({
             "name"       : name,
@@ -151,6 +146,9 @@
         # children.
         try:
             element = top["class"](*top["children"], **top["attributes"])
+            if top["class"] == WebDAVUnknownElement:
+                element.namespace = name[0]
+                element.name = name[1]
         except ValueError, e:
             e.args = ("%s at %s" % (e.args[0], self.location()),) + e.args[1:]
             raise # Re-raises modified e, but preserves traceback
