Index: twisted/web2/dav/http.py
===================================================================
--- twisted/web2/dav/http.py	(revision 18545)
+++ twisted/web2/dav/http.py	(working copy)
@@ -227,7 +227,7 @@
 
         if len(property.children) > 0:
             # Re-instantiate as empty element.
-            property = property.__class__()
+            property = property.cloneEmpty()
 
         if code > 400: # Error codes only
             log.err("Error during %s for %s: %s" % (self.method, property, message))
