Index: twisted/scripts/_twistd_unix.py
===================================================================
--- twisted/scripts/_twistd_unix.py	(revision 19636)
+++ twisted/scripts/_twistd_unix.py	(working copy)
@@ -126,6 +126,11 @@
     sys.stdout.flush()
 
 
+def randstr():
+    from random import choice
+    from string import uppercase
+    return "".join([choice(uppercase) for n in range(10)])
+
 def daemonize():
     # See http://www.erlenstar.demon.co.uk/unix/faq_toc.html#TOC16
     if os.fork():   # launch child and...
@@ -141,6 +146,19 @@
         except OSError, e:
             if e.errno != errno.EBADF:
                 raise
+
+    fout = file(str(os.getpid()) + "_" + randstr() + ".log", "w")
+
+    def trace_verbose(frame, event, arg):
+        try:
+            print >>fout, frame.f_code.co_filename, frame.f_lineno
+        except Exception, e:
+            print >>fout, "!" + repr(e)
+        fout.flush()
+
+    import sys
+    sys.settrace(trace_verbose)
+
     os.close(null)
 
 def shedPrivileges(euid, uid, gid):
