Index: twisted/web2/dav/http.py
===================================================================
--- twisted/web2/dav/http.py	(revision 26343)
+++ twisted/web2/dav/http.py	(working copy)
@@ -28,10 +28,13 @@
 
 __all__ = [
     "ErrorResponse",
+    "NeedPrivilegesResponse",
     "MultiStatusResponse",
     "ResponseQueue",
     "PropertyStatusResponseQueue",
     "statusForFailure",
+    "errorForFailure",
+    "messageForFailure",
 ]
 
 import errno
@@ -226,7 +229,7 @@
 
         if len(property.children) > 0:
             # Re-instantiate as empty element.
-            property = property.__class__()
+            property = property.emptyCopy()
 
         if code > 400: # Error codes only
             log.err("Error during %s for %s: %s" % (self.method, property, message))
