Index: twisted/web2/resource.py
===================================================================
--- twisted/web2/resource.py	(revision 19773)
+++ twisted/web2/resource.py	(working copy)
@@ -99,13 +99,13 @@
         response.headers.setHeader("allow", self.allowedMethods())
         return response
 
-    def http_TRACE(self, request):
-        """
-        Respond to a TRACE request.
-        @param request: the request to process.
-        @return: an object adaptable to L{iweb.IResponse}.
-        """
-        return server.doTrace(request)
+#    def http_TRACE(self, request):
+#        """
+#        Respond to a TRACE request.
+#        @param request: the request to process.
+#        @return: an object adaptable to L{iweb.IResponse}.
+#        """
+#        return server.doTrace(request)
 
     def http_HEAD(self, request):
         """
