Index: twisted/scripts/_twistd_unix.py
===================================================================
--- twisted/scripts/_twistd_unix.py	(revision 19636)
+++ twisted/scripts/_twistd_unix.py	(working copy)
@@ -9,7 +9,7 @@
 from twisted.scripts import mktap
 from twisted import copyright
 
-import os, errno, sys
+import os, errno, sys, time
 
 class ServerOptions(app.ServerOptions):
     synopsis = "Usage: twistd [options]"
@@ -129,9 +129,11 @@
 def daemonize():
     # See http://www.erlenstar.demon.co.uk/unix/faq_toc.html#TOC16
     if os.fork():   # launch child and...
+        time.sleep(2)
         os._exit(0) # kill off parent
     os.setsid()
     if os.fork():   # launch child and...
+        time.sleep(2)    
         os._exit(0) # kill off parent again.
     os.umask(077)
     null=os.open('/dev/null', os.O_RDWR)
